# Python NLTK: Texts and Frequencies

**(C) 2017-2023 by [Damir Cavar](http://damir.cavar.me/) <<dcavar@iu.edu>>**

**Version:** 0.9, January 2023

**Download:** This and various other Jupyter notebooks are available from my [GitHub repo](https://github.com/dcavar/python-tutorial-notebooks).

**License:** [Creative Commons Attribution-ShareAlike 4.0 International License](https://creativecommons.org/licenses/by-sa/4.0/) ([CA BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/))

This is a brief introduction to NLTK for simple frequency analysis of texts. I created this notebook for intro to corpus linguistics and natural language processing classes at Indiana University between 2017 and 2020.

For this to work, in the folder with the notebook we expect a subfolder data that contains a file HOPG.txt. This file contains the novel "A House of Pomegranates" by Oscar Wilde taken as raw text from [Project Gutenberg](https://www.gutenberg.org/).

## Simple File Processing

Reading a text into memory in Python is faily simple. We open a file, read from it, and close the file again. The following code prints out the first 300 characters of the text in memory:

In [1]:
ifile = open("data/HOPG.txt", mode='r', encoding='utf-8')
text = ifile.read()
ifile.close()
print(text[:300], "...")

A HOUSE OF POMEGRANATES




Contents:

The Young King
The Birthday of the Infanta
The Fisherman and his Soul
The Star-child




THE YOUNG KING




[TO MARGARET LADY BROOKE--THE RANEE OF SARAWAK]


It was the night before the day fixed for his coronation, and the
young King was sitting alone in his b ...


The optional parameters in the *open* function above define the **mode** of operations on the file and the **encoding** of the content. For example, setting the **mode** to **r** declares that *reading* from the file is the only permitted operation that we will perform in the following code. Setting the **encoding** to **utf-8** declares that all characters will be encoded using the [Unicode](https://en.wikipedia.org/wiki/Unicode) encoding schema [UTF-8](https://en.wikipedia.org/wiki/UTF-8) for the content of the file.

We can now import the [NLTK](https://www.nltk.org/) module in Python to work with frequency profiles and [n-grams](https://en.wikipedia.org/wiki/N-gram) using the tokens or words in the text.

In [2]:
import nltk

We can now lower the text, which means normalizing it to all characters lower case:

In [3]:
text = text.lower()
print(text[:300], "...")

a house of pomegranates




contents:

the young king
the birthday of the infanta
the fisherman and his soul
the star-child




the young king




[to margaret lady brooke--the ranee of sarawak]


it was the night before the day fixed for his coronation, and the
young king was sitting alone in his b ...


To generate a frequency profile from the text file, we can use the [NLTK](https://www.nltk.org/) function *FreqDist*:

In [4]:
myFD = nltk.FreqDist(text)

In [5]:
print(myFD)

<FreqDist with 39 samples and 174018 outcomes>


We can remove certain characters from the distribution, or alternatively replace these characters in the text variable. The following loop removes them from the frequency profile in myFD, which is a dictionary data structure in Python.

In [6]:
for x in ":,.-[];!'\"\t\n/ ?":
    del myFD[x]

We can print out the frequency profile by looping through the returned data structure:

In [7]:
for x in myFD:
    print(x, myFD[x])

e 17372
t 12521
a 11231
h 10802
o 9408
n 8843
i 8307
s 8093
r 7603
d 7249
l 5270
w 3665
m 3271
u 3269
f 3089
c 2693
g 2666
y 2168
p 1884
b 1812
v 1122
k 1026
j 103
q 81
z 61
x 48


To relativize the frequencies, we need to compute the total number of characters. This is assuming that we removed all punctuation symbols. The frequency distribution instance myFD provides a method to access the values associated with the individual characters. This will return a list of values, that is the frequencies associated with the characters.

In [8]:
myFD.values()

dict_values([11231, 10802, 9408, 3269, 8093, 17372, 3089, 1884, 3271, 2666, 7603, 8843, 12521, 2693, 2168, 1026, 8307, 1812, 7249, 5270, 3665, 48, 1122, 81, 103, 61])

The *sum* function can summarize these values in its list argument:

In [9]:
sum(myFD.values())

133657

To avoid type problems when we compute the relative frequency of characters, we can convert the total number of characters into a *float*. This will guarantee that the division in the following relativisation step will be a *float* as well.

In [10]:
float(sum(myFD.values()))

133657.0

We store the resulting number of characters in the *total* variable:

In [11]:
total = float(sum(myFD.values()))
print(total)

133657.0


We can now generate a probability distribution over characters. To convert the frequencies into relative frequencies we use list comprehension and divide every single list element by total. The resulting relative frequencies are stored in the variable *relfreq*:

In [12]:
relfrq = [ x/total for x in myFD.values() ]
print(relfrq)

[0.08402852076584093, 0.0808188123330615, 0.07038913038598801, 0.024458127894536014, 0.06055051362816762, 0.1299744869329702, 0.02311139708357961, 0.014095782488010355, 0.024473091570213306, 0.019946579677832064, 0.056884413087230745, 0.06616189200715264, 0.09368009157769515, 0.020148589299475522, 0.016220624434186013, 0.007676365622451499, 0.06215162692563801, 0.013557090163627793, 0.05423584249234982, 0.03942928540966803, 0.0274209356786401, 0.0003591282162550409, 0.008394622054961581, 0.0006060288649303815, 0.0007706292973806085, 0.0004563921081574478]


Let us compute the [Entropy](https://en.wikipedia.org/wiki/Entropy_(information_theory)) for the character distribution using the relative frequencies. We will need the [logarithm](https://en.wikipedia.org/wiki/Logarithm) function from the Python *math* module for that:

In [13]:
from math import log

We can define the [Entropy](https://en.wikipedia.org/wiki/Entropy_(information_theory)) function according to the equation $I = - \sum P(x) log_2( P(x) )$ as:

In [14]:
def entropy(p):
    return -sum( [ x * log(x, 2) for x in p ] )

In [15]:
entropy([1/8, 1/16, 1/4, 1/8, 1/16, 1/16, 1/4, 1/16])

2.75

In [16]:
entropy([1/8, 1/8, 1/8, 1/8, 1/8, 1/8, 1/8, 1/8])

3.0

We can now compute the entropy of the character distribution:

In [17]:
print(entropy([ 1/len(relfrq) ] * len(relfrq)))
print(entropy(relfrq))

4.700439718141092
4.124824125135032


We might be interested in the point-wise entropy of the characters in this distribution, thus needing the entropy of each single character. We can compute that in the following way:

In [18]:
entdist = [ -x * log(x, 2) for x in relfrq ]
print(entdist)

[0.3002319806578157, 0.29330480822209687, 0.2694850339615855, 0.13093762006835424, 0.24497024185626542, 0.38260584969385675, 0.12561626066788154, 0.08666922421352652, 0.13099613411450642, 0.11265259339509692, 0.2352638524022937, 0.2592127456210649, 0.32002184459468397, 0.11350057702872672, 0.09644826809662171, 0.053929238563860095, 0.24910770047189124, 0.08411915007238721, 0.2280405439219216, 0.18392139623527803, 0.1422756742696596, 0.004109580806277946, 0.05789199083219161, 0.006477433994507777, 0.007969598004767611, 0.005064783367910896]


We could now compute the variance over this point-wise entropy distribution or other properties of the frequency distribution as for example median, mode, or standard deviation.

## From Characters to Words/Tokens

We see that the frequency profile is for the characters in the text, not the words or tokens. In order to generate a frequency profile over words/tokens in the text, we need to utilize a **tokenizer**. [NLTK](https://www.nltk.org/) provides basic tokenization functions. We will use the *word_tokenize* function to generate a list of tokens:

In [19]:
tokens = nltk.word_tokenize(text)

We can print out the first 20 tokens to verify our data structure is a list with lower-case strings:

In [20]:
tokens[:20]

['a',
 'house',
 'of',
 'pomegranates',
 'contents',
 ':',
 'the',
 'young',
 'king',
 'the',
 'birthday',
 'of',
 'the',
 'infanta',
 'the',
 'fisherman',
 'and',
 'his',
 'soul',
 'the']

We can now generate a frequency profile from the token list, as we did with the characters above:

In [21]:
myTokenFD = nltk.FreqDist(tokens)
print(myTokenFD)

<FreqDist with 4009 samples and 38127 outcomes>


The frequency profile can be printed out in the same way as above by looping over the tokens and their frequencies. Note that we restrict the loop to the first 20 tokens here just to keep the notebook smaller. You can remove the [:20] selector in your own experiments.

In [22]:
print(tuple(myTokenFD.items())[:10])

(('a', 673), ('house', 26), ('of', 1108), ('pomegranates', 6), ('contents', 1), (':', 33), ('the', 2552), ('young', 112), ('king', 70), ('birthday', 8))


In [23]:
for token in list(myTokenFD.items()):
    print(token[0], token[1])

a 673
house 26
of 1108
pomegranates 6
contents 1
: 33
the 2552
young 112
king 70
birthday 8
infanta 37
fisherman 76
and 2118
his 483
soul 99
star-child 42
[ 4
to 697
margaret 1
lady 4
brooke 1
-- 26
ranee 1
sarawak 1
] 4
it 353
was 356
night 14
before 53
day 46
fixed 2
for 267
coronation 4
, 2675
sitting 2
alone 10
in 472
beautiful 25
chamber 10
. 1341
courtiers 3
had 235
all 82
taken 5
their 176
leave 10
him 372
bowing 2
heads 7
ground 16
according 1
ceremonious 1
usage 1
retired 3
great 83
hall 4
palace 25
receive 3
few 9
last 14
lessons 1
from 184
professor 1
etiquette 2
; 60
there 97
being 11
some 52
them 154
who 132
still 14
quite 21
natural 4
manners 1
which 59
courtier 1
is 242
i 376
need 13
hardly 4
say 11
very 16
grave 7
offence 1
lad 5
he 645
only 30
but 212
sixteen 1
years 16
age 4
not 208
sorry 1
at 203
departure 1
flung 12
himself 42
back 42
with 391
deep 13
sigh 1
relief 1
on 209
soft 5
cushions 1
embroidered 7
couch 8
lying 15
wild-eyed 1
open-mouthed 1
like 60
brown 6
w

In [25]:
stopwords = nltk.corpus.stopwords.words('english')
print(stopwords)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [26]:
for x in stopwords:
    del myTokenFD[x]
print(tuple(myTokenFD)[:20])

(',', '.', "'", 'said', 'thou', '?', 'thee', 'little', 'young', 'soul', 'came', 'great', 'answered', 'upon', 'one', 'fisherman', 'thy', 'gold', 'king', 'went')


## Counting N-grams

[NLTK](https://www.nltk.org/) provides simple methods to generate [n-gram](https://en.wikipedia.org/wiki/N-gram) models or frequency profiles over [n-grams](https://en.wikipedia.org/wiki/N-gram) from any kind of list or sequence. We can for example generate a bi-gram model, that is an [n-grams](https://en.wikipedia.org/wiki/N-gram) model for n = 2, from the text tokens:

In [27]:
myTokenBigrams = nltk.ngrams(tokens, 2)

To store the bigrams in a list that we want to process and analyze further, we convert the **Python generator object** myTokenBigrams to a list:

In [28]:
bigrams = list(myTokenBigrams)

Let us verify that the resulting data structure is indeed a list of string tuples. We will print out the first 20 tuples from the bigram list:

In [29]:
print(bigrams[:20])

[('a', 'house'), ('house', 'of'), ('of', 'pomegranates'), ('pomegranates', 'contents'), ('contents', ':'), (':', 'the'), ('the', 'young'), ('young', 'king'), ('king', 'the'), ('the', 'birthday'), ('birthday', 'of'), ('of', 'the'), ('the', 'infanta'), ('infanta', 'the'), ('the', 'fisherman'), ('fisherman', 'and'), ('and', 'his'), ('his', 'soul'), ('soul', 'the'), ('the', 'star-child')]


We can now verify the number of bigrams and check that there are exactly *number of tokens - 1 = number of bigrams* in the resulting list:

In [30]:
print(len(bigrams))
print(len(tokens))

38126
38127


The frequency profile from these bigrams is generated in exactly the same way as from the token list in the examples above:

In [31]:
myBigramFD = nltk.FreqDist(bigrams)
print(myBigramFD)

<FreqDist with 17766 samples and 38126 outcomes>


If we would want to know some more general properties of the frequency distribution, we can print out information about it. The print statement for this bigram frequency distribution tells us that we have 17,766 types and 38,126 tokens:

In [32]:
print(myBigramFD)

<FreqDist with 17766 samples and 38126 outcomes>


The bigrams and their corresponding frequencies can be printed using a *for* loop. We restrict the number of printed items to 20, just to keep this list reasonably long. If you would like to see the full frequency profile, remove the [:20] restrictor.

In [33]:
for bigram in list(myBigramFD.items())[:20]:
    print(bigram[0], bigram[1])
print("...")

('a', 'house') 4
('house', 'of') 5
('of', 'pomegranates') 4
('pomegranates', 'contents') 1
('contents', ':') 1
(':', 'the') 2
('the', 'young') 107
('young', 'king') 24
('king', 'the') 1
('the', 'birthday') 3
('birthday', 'of') 3
('of', 'the') 354
('the', 'infanta') 33
('infanta', 'the') 1
('the', 'fisherman') 5
('fisherman', 'and') 3
('and', 'his') 53
('his', 'soul') 51
('soul', 'the') 1
('the', 'star-child') 41
...


Pretty printing the bigrams is possible as well:

In [34]:
for ngram in list(myBigramFD.items()):
    print(" ".join(ngram[0]), ngram[1])
print("...")

a house 4
house of 5
of pomegranates 4
pomegranates contents 1
contents : 1
: the 2
the young 107
young king 24
king the 1
the birthday 3
birthday of 3
of the 354
the infanta 33
infanta the 1
the fisherman 5
fisherman and 3
and his 53
his soul 51
soul the 1
the star-child 41
star-child the 1
king [ 1
[ to 4
to margaret 1
margaret lady 1
lady brooke 1
brooke -- 1
-- the 4
the ranee 1
ranee of 1
of sarawak 1
sarawak ] 1
] it 2
it was 52
was the 20
the night 3
night before 1
before the 8
the day 7
day fixed 1
fixed for 1
for his 7
his coronation 2
coronation , 3
, and 1271
and the 287
king was 1
was sitting 1
sitting alone 1
alone in 1
in his 31
his beautiful 1
beautiful chamber 1
chamber . 1
. his 8
his courtiers 1
courtiers had 1
had all 2
all taken 1
taken their 1
their leave 1
leave of 1
of him 8
him , 146
, bowing 1
bowing their 1
their heads 5
heads to 1
to the 149
the ground 15
ground , 6
, according 1
according to 1
the ceremonious 1
ceremonious usage 1
usage of 1
day , 3
and had 

You can remove the [:20] restrictor above and print out the entire frequency profile. If you select and copy the profile to your clipboard, you can paste it into your favorite spreadsheet software and sort, analyze, and study the distribution in many interesting ways.

Instead of running the frequency profile through a loop we can also use a list comprehension construction in Python to generate a list of tuples with the n-gram and its frequency:

In [35]:
ngrams = [ (" ".join(ngram), myBigramFD[ngram]) for ngram in myBigramFD ]
print(ngrams[:100])

[(', and', 1271), ('of the', 354), ('and the', 287), ('. and', 205), (". '", 193), ('in the', 158), (", '", 155), ('to the', 149), ('him ,', 146), ('. the', 112), ("' and", 108), ('the young', 107), (', but', 96), ('and he', 89), (', for', 85), ("? '", 84), ('on the', 80), ('said to', 79), ('to him', 77), ('and said', 77), (', he', 73), ('and when', 71), ('young fisherman', 71), (', the', 70), ('from the', 69), ('. he', 64), ('he had', 63), ('through the', 58), ('into the', 55), ('at the', 55), ('i will', 55), ('the little', 54), ('and his', 53), ('it was', 52), ('his soul', 51), ('said ,', 49), ('of a', 48), ('out of', 48), ('that he', 47), ('and a', 47), ('when the', 47), ('for the', 47), ('in a', 46), ('with the', 45), ('it ,', 45), ('me ,', 45), ('. but', 44), ('him .', 44), ('he was', 43), ('with a', 43), ('when he', 43), ("' i", 42), ('it .', 42), ('the star-child', 41), ('of his', 41), ('i have', 41), ('her ,', 39), ('had been', 38), ('the sea', 38), ("' he", 37), ('by the', 36)

We can generate an increasing frequency profile using the sort function on the second element of the tuple list, that is on the frequency:

In [36]:
sortedngrams = sorted(ngrams, key=lambda x: x[1])
print(sortedngrams[:20])
print("...")

[('pomegranates contents', 1), ('contents :', 1), ('king the', 1), ('infanta the', 1), ('soul the', 1), ('star-child the', 1), ('king [', 1), ('to margaret', 1), ('margaret lady', 1), ('lady brooke', 1), ('brooke --', 1), ('the ranee', 1), ('ranee of', 1), ('of sarawak', 1), ('sarawak ]', 1), ('night before', 1), ('day fixed', 1), ('fixed for', 1), ('king was', 1), ('was sitting', 1)]
...


We can increase the speed of this *sorted* call by using the *itemgetter()* function in the *operator* module. Let us import this function:

In [37]:
from operator import itemgetter

We can now define the sort-key for *sorted* using the *itemgetter* function and selecting with 1 the second element in the tuple. Remember that the enumeration of elements in lists or tuples in Python starts at 0.

In [38]:
sortedngrams = sorted(ngrams, key=itemgetter(1))
print(sortedngrams[:20])
print("...")

[('pomegranates contents', 1), ('contents :', 1), ('king the', 1), ('infanta the', 1), ('soul the', 1), ('star-child the', 1), ('king [', 1), ('to margaret', 1), ('margaret lady', 1), ('lady brooke', 1), ('brooke --', 1), ('the ranee', 1), ('ranee of', 1), ('of sarawak', 1), ('sarawak ]', 1), ('night before', 1), ('day fixed', 1), ('fixed for', 1), ('king was', 1), ('was sitting', 1)]
...


A decreasing frequency profile can be generated using another parameter to *sorted*:

In [39]:
sortedngrams = sorted(ngrams, key=itemgetter(1), reverse=True)
print(sortedngrams[:20])
print("...")

[(', and', 1271), ('of the', 354), ('and the', 287), ('. and', 205), (". '", 193), ('in the', 158), (", '", 155), ('to the', 149), ('him ,', 146), ('. the', 112), ("' and", 108), ('the young', 107), (', but', 96), ('and he', 89), (', for', 85), ("? '", 84), ('on the', 80), ('said to', 79), ('to him', 77), ('and said', 77)]
...


We can pretty-print the decreasing frequency profile:

In [40]:
sortedngrams = sorted(ngrams, key=itemgetter(1), reverse=True)
for t in sortedngrams[:20]:
    print(t[0], t[1])
print("...")

, and 1271
of the 354
and the 287
. and 205
. ' 193
in the 158
, ' 155
to the 149
him , 146
. the 112
' and 108
the young 107
, but 96
and he 89
, for 85
? ' 84
on the 80
said to 79
to him 77
and said 77
...


In [41]:
total = float(sum(myBigramFD.values()))
exceptions = ["]", "[", "--", ",", ".", "'s", "?", "!", "'", "'ye"]
results = []
for x in myBigramFD:
    if x[0] in exceptions or x[1] in exceptions:
        continue
    if x[0] in stopwords or x[1] in stopwords:
        continue
    results.append( (x[0], x[1], myBigramFD[x]/total) )
sortedresults = sorted(results, key=itemgetter(2), reverse=True)
for x in sortedresults[:20]:
    print(x[0], x[1], x[2])

young fisherman 0.0018622462361642972
young king 0.0006294916854639878
thou hast 0.0005508052247809893
thou art 0.0005245764045533231
little dwarf 0.000498347584325657
thou shalt 0.00034097466295966006
little mermaid 0.0003147458427319939
soul answered 0.0003147458427319939
give thee 0.0002622882022766616
let us 0.0002622882022766616
get thee 0.00023605938204899544
soul said 0.00023605938204899544
yellow gold 0.00023605938204899544
whole world 0.00020983056182132927
art thou 0.00020983056182132927
day long 0.00020983056182132927
'thou hast 0.00020983056182132927
thee gone 0.00020983056182132927
white rose 0.00020983056182132927
white gold 0.00020983056182132927


To be continued...

(C) 2017-2023 by [Damir Cavar](http://damir.cavar.me/) <<dcavar@iu.edu>>